zkvm-host/bin/agg.rs

```diff
@@ -0,0 +1,123 @@
+use std::fs;
+
+use anyhow::Result;
+use cargo_metadata::MetadataCommand;
+use clap::Parser;
+use client_utils::{types::AggregationInputs, RawBootInfo, BOOT_INFO_SIZE};
+use host_utils::fetcher::{ChainMode, SP1KonaDataFetcher};
+use sp1_sdk::{utils, ProverClient, SP1Proof, SP1ProofWithPublicValues, SP1Stdin};
+
+pub const AGG_ELF: &[u8] = include_bytes!("../../elf/aggregation-client-elf");
+pub const MULTI_BLOCK_ELF: &[u8] = include_bytes!("../../elf/validity-client-elf");
+
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Start L2 block number.
+    #[arg(short, long, num_args = 1.., value_delimiter = ',')]
+    proofs: Vec<String>,
+
+    #[arg(short, long)]
+    multi: bool,
+
+    /// Verbosity level.
+    #[arg(short, long, default_value = "0")]
+    verbosity: u8,
+}
+
+/// Load the aggregation proof data.
+fn load_aggregation_proof_data(proof_names: Vec<String>) -> (Vec<SP1Proof>, Vec<RawBootInfo>) {
+    let metadata = MetadataCommand::new().exec().unwrap();
+    let workspace_root = metadata.workspace_root;
+    let proof_directory = format!("{}/data/proofs", workspace_root);
+
+    let mut proofs = Vec::with_capacity(proof_names.len());
+    let mut boot_infos = Vec::with_capacity(proof_names.len());
+
+    for proof_name in proof_names.iter() {
+        let proof_path = format!("{}/{}.bin", proof_directory, proof_name);
+        if fs::metadata(&proof_path).is_err() {
+            panic!("Proof file not found: {}", proof_path);
+        }
+        let mut deserialized_proof =
+            SP1ProofWithPublicValues::load(proof_path).expect("loading proof failed");
+        proofs.push(deserialized_proof.proof);
+
+        // The public values are the ABI-encoded BootInfo.
+        let mut boot_info_buf = [0u8; BOOT_INFO_SIZE];
+        deserialized_proof
+            .public_values
+            .read_slice(&mut boot_info_buf);
+        let boot_info = RawBootInfo::abi_decode(&boot_info_buf).unwrap();
+        boot_infos.push(boot_info);
+    }
+
+    (proofs, boot_infos)
+}
+
+// Execute the Kona program for a single block.
+#[tokio::main]
+async fn main() -> Result<()> {
+    utils::setup_logger();
+
+    let args = Args::parse();
+    let fetcher = SP1KonaDataFetcher::new();
+    let prover = ProverClient::new();
+
+    let (proofs, boot_infos) = load_aggregation_proof_data(args.proofs);
+
+    // Fetch the headers from the L1 head of the first boot info to the current L1 head, inclusive.
+    let first_head = boot_infos.last().unwrap().l1_head;
+
+    // Confirm that the headers are in the correct order.
+    let start_header = fetcher
+        .get_header_by_hash(ChainMode::L1, first_head)
+        .await?;
+    // End header is the current L1 head.
+    // TODO: Should this be the finalized header/fetched from a different source?
+    let end_header = fetcher.get_head(ChainMode::L1).await?;
```

Comment:
@ratankaliani Yeah, this needs to be passed in as a part of the call to this function. Basically `op-proposer` will take its view of the current head, checkpoint it (it will have progressed by the time the checkpoint is done), and then pass that L1 Head to its call to the server to make sure the same one is used in the proof.

Reply:
We can pass the hash and then do `get_header_by_hash()`.